streamer/streamer.py

```diff
@@ -285,139 +306,172 @@ def get_indexed_files(product, year=None, month=None, datacube_env=None):
+        with open(config) as cfg_file:
             cfg = yaml.load(cfg_file)
     else:
         cfg = yaml.load(DEFAULT_CONFIG)
 
-    product_config = cfg['products'][product]
-
-    cog_convert = COGNetCDF(**product_config)
-
-    if flist is not None:
-        with open(flist, 'r') as fb:
+    try:
+        with open(filelist) as fb:
             file_list = np.genfromtxt(fb, dtype='str')
+        tasks = file_list.size
+    except FileNotFoundError:
+        LOG.error(f'MPI Worker ({MPI_JOB_RANK}): No netCDF file/s found in the input path')
+        raise
     else:
-        file_list = list(filenames)
+        if tasks == 0:
+            _raise_value_err(f'MPI Worker ({MPI_JOB_RANK}): No netCDF file/s found in the input path')
 
-    try:
-        comm = MPI.Comm.Get_parent()
-        size = comm.Get_size()
-        rank = comm.Get_rank()
-    except:
-        LOG.info("Run with single process")
-        for filename in file_list:
-            cog_convert(filename, output_dir)
+    product_config = cfg['products'][product]
+    num_workers = numprocs if numprocs > 0 else _raise_value_err(
+        f"MPI Worker ({MPI_JOB_RANK}): Number of processes cannot be zero")
+
+    # Ensure all errors/exceptions are handled before this, else master-worker processes
+    # will enter a dead-lock situation
+    if MPI_JOB_RANK == 0:
+        name = MPI.Get_processor_name()
+        task_index = 0
+        closed_workers = 0
+        job_args = []
+        LOG.debug(f"MPI Master ({MPI_JOB_RANK}) on {name} node, starting with {num_workers} workers")
+
+        # Append the jobs_args list for each filename to be scheduled among all the available workers
+        if tasks == 1:
+            job_args = [(product_config, str(file_list), output_dir)]
+        elif tasks > 1:
+            for filename in file_list:
+                job_args.extend([(product_config, str(filename), output_dir)])
+
+        while closed_workers < num_workers:
+            MPI_COMM.recv(source=MPI.ANY_SOURCE, tag=MPI.ANY_TAG, status=MPI_JOB_STATUS)
+            source = MPI_JOB_STATUS.Get_source()
+            tag = MPI_JOB_STATUS.Get_tag()
+
+            if tag == TagStatus.READY:
+                # Worker is ready, so assign a task
+                if task_index < tasks:
+                    MPI_COMM.send(job_args[task_index], dest=source, tag=TagStatus.START)
+                    LOG.debug("MPI Master (%d) assigning task to worker (%d): Process %r file" % (MPI_JOB_RANK,
+                                                                                                  source,
+                                                                                                  job_args[task_index]))
+                    task_index += 1
+                else:
+                    MPI_COMM.send(None, dest=source, tag=TagStatus.EXIT)
+            elif tag == TagStatus.DONE:
+                LOG.debug(f"MPI Worker ({source}) on {name} completed the assigned task")
```

Comment:
`status.Done` is a bit redundant to me. The worker can go back to `status.Ready` straight away.

Reply:
Hi @emmaai, `status.Done` was added to aid debugging and better logging. And since `LOG` is configured as `StreamHandler(sys.stdout)` only, conflicts between processes would not arise.

I will implement the option to run without MPI in my next PR.
Hope you are fine with these changes.